contributor-docs/behaviors.md

```diff
@@ -0,0 +1,132 @@
+
+### Examples
+
+* `useProvidedRefOrCreate`
+* `usePosition`
+* `useClickAway`
+* `useTypeAhead`
+
+## Component behaviors
+
+Component behaviors specifically implement behaviors for components. Therefore, hooks that implement component behaviors should always return an object containing props objects to be spread across the elements that make up the component's JSX. While component behaviors do not necessarily have to be one-to-one, **a component may only use one component behavior**. Sometimes it makes sense to use the same component behavior on several component. For example: the Dialog, Popover, and PointerBox components may all use the `usePopover` component behavior.
+
+### Examples
+
+* `usePopover`
+* `useComboBox`
+* `useFilteredSearch`
+* `useDetails`
+
+## Implementation guidelines
+
+All behaviors should be implemented as [React Hooks](https://reactjs.org/docs/hooks-intro.html).
+
+Component behaviors should be relatively brief. A component behavior should build a component's experience through the combination of several configured generic behaviors.
+
+Example: The Dialog Component uses a `usePopover` behavior. The `usePopover` behavior in turn may call generic behaviors such as `usePosition` (to position the popover on the screen), `useClickAway` (to set up an event handler for clicking outside of the dialog), and `useFocusTrap` (to prevent focus from leaving the dialog while it is being shown).
+
+### Component implementation
+
+Component implementation is [discussed elsewhere](https://github.com/primer/components/blob/main/contributor-docs/CONTRIBUTING.md#developing-components), but with respect to behaviors, follow these guidelines.
+
+* If a component has behaviors, it should make a call to exactly one component behavior hook.
+* The hook will return an object containing multiple props objects. Each of the returned objects are props that should be spread onto the appropriate JSX element.
+* Component implementations should be as clean as possible, mostly consisting of the JSX tree definition that the component renders.
+* Component props and, in some circumstances, context values, should serve as arguments to component behavior hooks.
+
+### [YAGNI Principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)
+
+As a library, there is a need to balance the YAGNI Principle with building a library that functions for a wide range of use cases. Each component and behavior will implement features on a case-by-base basis. When you are adding a feature that seems too specific, _generalize_ it. Can you account for multiple use cases by implementing a single feature? Can you take a single use case and identify a feature that would enable this use case along with others?
```

Comment:
I've gone back and forth on my thoughts on this section -  I think YAGNI will mean different things for different people as far as what it looks like in practice _and_ in the past in PRC we haven't done a great job of anticipating a wide enough array of use cases for components during the authoring process which has led to some unnecessary breaking changes that may have been avoided if we had taken a less lean approach up front, did more auditing of current use cases in .com, etc. There's probably a spectrum of how ruthless teams should be with the YAGNI principle depending on what kind of software they are writing, and a systems team might be on the lighter end. Maybe just renaming this section `Balance API surface area with anticipating the needs of the system`? Or something more generic like that would help?

Also, I would expand on what `Each component and behavior will implement features on a case-by-base basis` means or remove it. Does that sentence mean we will add features one at a time as they are needed? Or does that mean each component will need to have research done to determine the right amount of features?

Reply:
I agree - great suggestions. I will update this section. I think we're aligned on this though:

> There's probably a spectrum of how ruthless teams should be with the YAGNI principle depending on what kind of software they are writing, and a systems team might be on the lighter end. 

Totally agree. I was trying to say something to this effect, but re-reading my paragraph here I think I could have been more clear.

Reply:
@emplums I have rewritten this section. Let me know if it makes more sense.